golang_src/datastore.go

```diff
@@ -0,0 +1,268 @@
+	var toRecipients, ccRecipients []string
+
+	for _, recipient := range e.Recipients {
+		switch recipient.Type {
+		case RecipientTo:
+			toRecipients = append(toRecipients, recipient.Address)
+		case RecipientCC:
+			ccRecipients = append(ccRecipients, recipient.Address)
+		case RecipientBCC:
+		}
+	}
+	if len(toRecipients) > 0 {
+		bodyBytes.WriteString(fmt.Sprintf("To: %s\r\n", strings.Join(toRecipients, ", ")))
+	}
+	if len(ccRecipients) > 0 {
+		bodyBytes.WriteString(fmt.Sprintf("Cc: %s\r\n", strings.Join(ccRecipients, ", ")))
+	}
+	bodyBytes.WriteString(fmt.Sprintf("Subject: %s\r\n", e.Subject))
+	bodyBytes.WriteString("\r\n")
+	bodyBytes.WriteString(e.Body)
+	return bodyBytes.Bytes()
+}
+
+func (m *MongoClient) dial() (*mgo.Session, error) {
+	hostPort := fmt.Sprintf("%s:%s", m.Config.Host, m.Config.Port)
+	timeout := time.Duration(m.Config.ConnectionTimeout) * time.Second
+	session, err := mgo.DialWithTimeout(hostPort, timeout)
+	if err != nil {
+		return nil, err
+	}
+	return session, nil
+}
+
+func (m *MongoClient) CleanUp() {
+	if m.session != nil {
+		m.session.Close()
+	}
+}
+
+func (m *MongoClient) getClient() (*mgo.Session, error) {
+	var connectionErr error
+
+	for i := 0; i < m.Config.ConnectionRetries; i++ {
+		if i > 0 {
+			log.Println(fmt.Sprintf("Failed to dial datastore with %d timeout, %d tries remaining", m.Config.ConnectionTimeout, m.Config.ConnectionRetries-i-1))
+		}
+		if m.session == nil {
+			if sess, err := m.dial(); err != nil {
+				connectionErr = err
+				continue
+			} else {
+				m.session = sess
+				m.session.SetSocketTimeout(time.Duration(m.Config.ConnectionTimeout) * time.Second)
+				m.session.SetSyncTimeout(time.Duration(m.Config.ConnectionTimeout) * time.Second)
+			}
+		}
+		if m.session != nil {
+			if err := m.session.Ping(); err != nil {
+				connectionErr = err
+				m.session = nil
+				continue
+			} else {
+				return m.session.Clone(), nil
+			}
+		}
+	}
+	// Couldn't access the datastore, what do we do?
+	// In a production system, this would (at least) send alerts/page people, because this
+	// may be lost data. We'll settle for a retry with backoff here, and fall back
+	// to an error.
+	return nil, errors.New(fmt.Sprintf("Failed to connect to the datastore. Error: %s", connectionErr.Error()))
+}
+
+func (m *MongoClient) FetchReadyEmail(workerId string) (*Email, error) {
+	session, err := m.getClient()
+	if err != nil {
+		return nil, err
+	}
+	session.SetMode(mgo.Strong, true)
+	emailCollection := session.DB(m.Config.DatabaseName).C("emails")
+	email := Email{}
+	change := mgo.Change{
+		Update:    bson.M{"$set": bson.M{"worker_id": workerId}},
+		ReturnNew: true}
+
+	// Fetch emails that are incomplete and N seconds old or older
+	olderThan := bson.M{"$lt": time.Now().Add(-time.Duration(m.Config.SendRetryInterval) * time.Second)}
+	_, err = emailCollection.Find(bson.M{"worker_id": nil, "status": "incomplete", "updated_at": olderThan}).Apply(change, &email)
```

Comment:
Yeah, good catch. In my original design doc I wanted to add a heart beat to the goroutines so the parent (or perhaps a janitor routine of sorts) could check the updated_at time on the Emails and reclaim them if a certain interval had expired. I decided to punt on that as I'd already spent so much time getting this to the point where I was ready to submit it.

I can take a stab at something like that though, if you'd like. I can certainly see the need

Reply:
I thought about this a little more after my original reply, and I figured it would be useful to see my thought process. Let me know what you think.

If a goroutine panics, it takes the whole process down, so that would lock up all the claimed emails, not just the one that dies.

The parent process knows which UUIDs it's assigning to the workers, so it could conceivably launch a cleanup goroutine that looks for the status of workers+emails in the database.

It could either:
a) look for "claimed" emails with an updated_at time >= $SOME_LARGE_INTERVAL and then reset those emails to worker_id == null. We'd also want to alert. Meanwhile the email update code should be amended to do something like a Compare-and-Swap when updating an email, and only update the record in the event it still has a claim to the email in question (in the event the goroutine was merely hung and then times out and recovers)

OR

b) Workers would periodically heartbeat and check in *AND* they would check in when they claim an email, or release one. When the cleanup routine sees a goroutine that hasn't sent a heartbeat in $SOME_INTERVAL seconds, then send an alert. All the emails in the database owned by that goroutine are marked as worker_id = null. We'd still need to update the logic to something more Compare-and-Swap like.

We can't safely just do something like reclaim all the emails with worker_ids we don't know; there may be other processes running that own those emails.

Also problematic: we can't really kill a goroutine from the outside. We can pass it a channel and tell it to quit, but if a socket is hung up trying to do something, telling it to quit via the channel could similarly be blocked. As such, it may in fact be superior to do a multi-worker model with monitoring that can kill and restart processes at will.

Reply:
These are good points! What if came up with some algo that would allow us to claim a message using a different field? That would eliminate the need for a `worker_id`.

Reply:
I'm writing something up and I'll paste it here when I'm done. I promise I won't ignore this 😄

Reply:
I played around with a couple different ideas but I kept coming back to something like this every time. I like the simplicity, and it's similar to an algorithm I implemented in a production cloud some years ago.

```
# We could use the existing status field and add a fourth status "active" for when they're claimed and we're trying to send them
  # Introduce a new state "queued" -> ("active", "incomplete", "failed", "complete")

  - Worker fetches one email from the databse where status = "incomplete", updated_at >= $SOME_INTERVAL_BEFORE_NOW before now and sets it status = "active", updated_at = now()
  - If there are no emails the worker goes to sleep
  - Else, Try to send the email
    - If the email is sent, mark it as "complete"
    - If it fails:
      - If non-recoverable or tries remaining, mark it as "failed"
      - If tries remaining, mark it as "ready" and increment "tries"
  - Meanwhile, we have a cleanup process that finds emails that are "active", updated_at >= $SOME_INTERVAL_BEFORE_NOW:
    - for each email found:
      - if tries remaining, mark as "incomplete", tries += 1
      - if no tries remaining, mark as "failed", tries = $MAX_TRIES
```

My other primary idea involved a process that populates a queue with emails ready to send, workers that pop them off, and then writes the status back to the database. However, this seemed needlessly complex for this implementation and there are other limitations, such as messages sitting in the queue so long that another process erroneously reclaims it and queues it up behind the original, perhaps endlessly

Reply:
Yeah, that sounds like a solid proposal. As far as I see no worker id is needed after all.

Reply:
Yup, agreed.